packages/modules/psa/api.py

```diff
@@ -0,0 +1,157 @@
+
+    # Filter list for given VIN or select first entry if no VIN is provided
+    try:
+        vehicle_selected = next(
+            vehicle for vehicle in vehicles if vehicle['vin'] == vehicle_vin) if vehicle_vin else vehicles[0]
+        vehicle_id = vehicle_selected['id']
+        vehicle_vin_found = vehicle_selected['vin']
+    except Exception:
+        log.error(f'Cannot find VIN {vehicle_vin} in PSA account {user_id}. Check VIN and retry.')
+        return
+
+    log.debug(f'Fetching details for VIN: {vehicle_vin_found} with vehicle_id:{vehicle_id}')
+
+    # Step 3: get details for selected vehicle
+    data = {'client_id': client_id}
+    headers = {
+        'Accept': 'application/hal+json', 'Authorization': 'Bearer %s' % access_token, 'x-introspect-realm': realm}
+    battery = req.get_http_session().get(
+        f'https://api.groupe-psa.com/connectedcar/v4/user/vehicles/{vehicle_id}/status?client_id={client_id}',
+        headers=headers).json()
+
+    # filter to only include type=Electric but remove all others. Seen type=Fuel and type=Electric being returned.
+    energies = battery['energy']
+    energy_selected = next(energy for energy in energies if energy['type'] == 'Electric')
+    soc = energy_selected['level']
+    range = energy_selected['autonomy']
+    soc_timestamp = energy_selected['updatedAt']
+
+# comment the manual calculation section for now
+#    if (int(soccalc) == 0):
+#        # manual calculation not enabled, using existing logic
+#        if (int(str(vehicle)) == 1):
+#            f = open('/var/www/html/openWB/ramdisk/soc', 'w')
+#        if (int(str(vehicle)) == 2):
+#            f = open('/var/www/html/openWB/ramdisk/soc1', 'w')
+#        f.write(str(soc))
+#        f.close()
+#    else:
+#        # manual calculation  enabled, using new logic with timestamp
+#        if (int(str(vehicle)) == 1):
+#            f = open('/var/www/html/openWB/ramdisk/psasoc', 'w')
+#        if (int(str(vehicle)) == 2):
+#            f = open('/var/www/html/openWB/ramdisk/psasoc1', 'w')
+#        f.write(str(soc))
+#        f.close()
+#        # getting timestamp of fetched SoC
+#        fetchedsoctime = batt[0]['updatedAt']
+#        soct = time.strptime(fetchedsoctime, "%Y-%m-%dT%H:%M:%SZ")
+#        soctime = time.mktime(soct)
+#        # adding one hour to UTC to get CET
+#        soctime = soctime + 3600
+#        # checking for daylight saving time
+#        dst = time.localtime()
+#        if (dst.tm_isdst == 1):
+#            # adding one hour to fetched SoCtime in daylight saving time
+#            soctime = soctime + 3600
+#        # writing timestamp to ramdisk
+#        if (int(str(vehicle)) == 1):
+#            f = open('/var/www/html/openWB/ramdisk/psasoctime', 'w')
+#        if (int(str(vehicle)) == 2):
+#            f = open('/var/www/html/openWB/ramdisk/psasoctime1', 'w')
+#        f.write(str(int(soctime)))
+#        f.close()
+    return soc, range, soc_timestamp
```

Comment:
Es könnte einfacher sein hier ein `CarState` zu erzeugen und zurückzugeben.

Reply:
das ist in den anderen Modulen ähnlich implementiert. Bin aber bei dir, das wäre schöner. Was mir leider nicht klar ist, wie kann ich den CarState in den anderen Funktionen, z.b. fetch_soc und in der soc.py direkt weiterverwenden ohne mir stets ein neues Objekt über carstate=CarState(soc, range, timestamp) zu erzeugen - das CarState Objekt muss hier ja im Grunde nur 1-1 durchgereicht werden. von PSARequest zu fetch_soc zur update in soc.py.